compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/draganddrop/cupertino/DragAndDrop.uikit.kt

```diff
@@ -0,0 +1,377 @@
+            doesSessionAllowMoveOperation(session)
+        }
+    }
+
+    private fun <R> delegating(block: DragSource.() -> R): R = dragSource.run(block)
+}
+
+private fun DragSource.toInteractionDelegate(): UIDragInteractionDelegateProtocol =
+    DragSourceToInteractionDelegateAdapter(this)
+
+
+@ExperimentalComposeUiApi
+interface DropTarget {
+    /**
+     * https://developer.apple.com/documentation/uikit/uidropinteractiondelegate/2891010-dropinteraction
+     */
+    fun UIDropInteraction.canHandleSession(session: UIDropSessionProtocol): Boolean
+
+    /**
+     * https://developer.apple.com/documentation/uikit/uidropinteractiondelegate/2890889-dropinteraction
+     */
+    fun UIDropInteraction.performDropFromSession(session: UIDropSessionProtocol)
+
+    /**
+     * https://developer.apple.com/documentation/uikit/uidropinteractiondelegate/2890888-dropinteraction
+     */
+    fun UIDropInteraction.proposalForSessionUpdate(session: UIDropSessionProtocol): UIDropProposal
+}
+
+/**
+ * A bridging class that adapts [DropTarget] to [UIDropInteractionDelegateProtocol].
+ */
+private class DropTargetToInteractionDelegateAdapter(
+    private val dropTarget: DropTarget
+) : CMPDropInteractionProxy() {
+
+    override fun canHandleSession(
+        session: UIDropSessionProtocol,
+        interaction: UIDropInteraction
+    ): Boolean = delegating {
+        with(interaction) {
+            canHandleSession(session)
+        }
+    }
+
+    override fun performDropFromSession(
+        session: UIDropSessionProtocol,
+        interaction: UIDropInteraction
+    ): Unit = delegating {
+        with(interaction) {
+            performDropFromSession(session)
+        }
+    }
+
+    override fun proposalForSessionUpdate(
+        session: UIDropSessionProtocol,
+        interaction: UIDropInteraction
+    ): UIDropProposal = delegating {
+        with(interaction) {
+            proposalForSessionUpdate(session)
+        }
+    }
+
+    private fun <R> delegating(block: DropTarget.() -> R): R = dropTarget.run(block)
+}
+
+private fun DropTarget.toInteractionDelegate(): UIDropInteractionDelegateProtocol =
+    DropTargetToInteractionDelegateAdapter(this)
+
+/**
+ * Modifier that adds an invisible view to the hierarchy that can be used as a drag-and-drop source
+ * or target.
+ *
+ * On iOS  drag-and-drop sessions can only be triggered by internal API of native views via adding
+ * [UIDragInteraction] and [UIDropInteraction]. It's not allowed to start a drag-and-drop session
+ * imperatively.
+ *
+ * This modifier allows to add a drag-and-drop session to a view and handle the drop event.
+ * Compose merely controls the position and size of the dummy views and forwards events of the
+ * drag-n-drop session via [dragSource] and [dropTarget].
+ */
+@ExperimentalComposeUiApi
+fun Modifier.dragAndDrop(
+    dragSource: DragSource? = null,
+    dropTarget: DropTarget? = null
+): Modifier {
+    val composedDragAndDropModifier = composed {
```

Comment:
Google are moving away from `Modifier.composed`. Can we implement this via `Modifier.Node`? You can access locals by subclassing `LayoutNode` and accessing `compositionLocalMap`.